ls1intum/Artemis
# [Feature] Structured Grading Instructions for Programming Exercises

### Checklist

### Motivation and Context


Programming exercises did not feature the structured grading instructions.

### Description

- Replaced only problem statement with  `AssessmentInstructionsComponent`
- Added SGI functionality to programming exercises assessment
- Added link to solution repo for programming exercises in example solution section of `AssessmentInstructionsComponent`
- Added SGI view to `ProgrammingExerciseDetailsComponent`

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Create programming exercise with due date and manual assessment
3.1 Now a toggle under grading instructions should be visible, toggle it and adapt the structured grading instructions
4. View programming exercise details and see if SGI are displayed
5. Participate in programming exercise
6. Assess exercise and check out the new instructions layout on the right side. It should contain Problem Statement, Example Solution, Grading Instruction
7. Add a inline feedback and drag and drop the structured grading instruction component to the inline feedback component
8. Inline feedback component should adapt accordingly 


### Test Coverage

* code-editor-tutor-assessment-inline-feedback.component.ts: 92%
* ProgrammingExerciseParticipationResource.java: 89%
* ProgrammingExerciseResource.java : 86%

### Screenshots



SGI toggle:
<img width="1185" alt="image" src="https://user-images.githubusercontent.com/41108487/96338381-e39bbe00-108d-11eb-81a2-285ee72f1515.png">


Instructions on the right side:
<img width="1411" alt="image" src="https://user-images.githubusercontent.com/41108487/96333898-f81b8e80-106c-11eb-8c22-3256fb8dbdf1.png">

SGI before DnD:
<img width="1417" alt="image" src="https://user-images.githubusercontent.com/41108487/96333943-4466ce80-106d-11eb-9e33-4cc49f1bff2e.png">

After DnD:
<img width="1386" alt="image" src="https://user-images.githubusercontent.com/41108487/96333946-4df03680-106d-11eb-903b-c79ef1ce9883.png">

ProgrammingExerciseDetails view:
<img width="1184" alt="image" src="https://user-images.githubusercontent.com/41108487/96338213-bd295300-108c-11eb-9ae4-c5e30cff1169.png">
